federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+	cachedService, ok := serviceCache.get(key)
+	if !ok {
+		// here we filtered all non-federation services
+		return nil
+	}
+	obj, exists, err := clusterCache.serviceStore.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		clusterCache.serviceQueue.Add(key)
+		return err
+	}
+	var needUpdate bool
+	if exists {
+		service, ok := obj.(*v1.Service)
+		if ok {
+			glog.V(4).Infof("Found service for federation service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+			needUpdate = cc.processServiceUpdate(cachedService, service, clusterName)
+		} else {
+			tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				return fmt.Errorf("object contained wasn't a service or a deleted key: %+v", obj)
+			}
+			glog.Infof("Found tombstone for %v", key)
+			needUpdate = cc.processServiceDeletion(cachedService, tombstone.Key, clusterName)
+		}
+	} else {
+		// service absence in store means watcher caught the deletion
+		glog.Infof("service has been deleted %v", key)
+		needUpdate = cc.processServiceDeletion(cachedService, key, clusterName)
+	}
+	if needUpdate {
+		cc.persistFedServiceUpdate(cachedService.lastState, fedClient)
+		cachedService.appliedState = cachedService.lastState
+	}
+	return nil
+}
+
+
+func (cc *clusterClientCache) processServiceDeletion(cachedService *cachedService, obj interface{}, clusterName string) bool{
+	glog.V(4).Infof("Processing service update for %s, cluster %s", obj, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	cachedStatus, ok := cachedService.serviceStatusMap[clusterName]
+	// cached status found, remove ingress info from federation service
+	if ok {
+		// new status = cached federation status - cached status for the cluster + new status from the cluster
+		cachedFedServiceStatus := cachedService.lastState.Status.LoadBalancer
+		for i, fed := range cachedFedServiceStatus.Ingress {
+			for _, old := range cachedStatus.Ingress{
+				// remove if same ingress record found
+				if (old.IP == fed.IP && old.Hostname == fed.Hostname) {
+					cachedFedServiceStatus.Ingress = append(cachedFedServiceStatus.Ingress[:i], cachedFedServiceStatus.Ingress[i+1:]...)
+					glog.V(4).Infof("remove old ingress info '%s', '%s' for service %s/%s", old.Hostname, old.IP, cachedService.lastState.Namespace, cachedService.lastState.Name)
+				}
+			}
+		}
+		delete(cachedService.serviceStatusMap, clusterName)
+		delete(cachedService.endpointMap, clusterName)
+		return true
+	} else {
+		glog.V(4).Infof("service removal %s/%s from cluster %s observed, but cached endpoint info is not changed.", cachedService.lastState.Namespace, cachedService.lastState.Name, clusterName)
+	}
+	return false
+}
+
+// Update ingress info when service updated, and persist to federation apiserver
+func (cc *clusterClientCache) processServiceUpdate(cachedService *cachedService, service *v1.Service, clusterName string) bool {
```

Comment:
nit: Start the comment with the function name.  godoc convention.
nit: Add comment to explain that return value indicates whether or not an update is required.